src/employees.service.ts

```diff
@@ -0,0 +1,147 @@
+        if (!childrenArray ) {
+            return;
+        }
+        for (let obj of childrenArray) {
+            let blank = {};
+            blank[parentField] = parentId;
+            result.push(Object.assign(blank, obj));
+            visitChildren(obj[containsField], obj[idField]);
+        }
+    }
+    visitChildren([data[0]], null);
+    return result;
+};
+
+export interface Employee {
+    id: number;
+    managerId: number;
+    name: string;
+    firstName: string;
+    reports: any;
+    lastName: string;
+    title: string;
+    active: boolean;
+    phone: number;
+    hireDate: Date;
+}
+export function createEmployee(id: number, managerId: number, firstName: string, lastName: string, title: string, phone: number, active: boolean = true, reports: any = [],  hireDate: Date = null): Employee {
+    const emp: Employee = {
+        id: id,
+        managerId: managerId,
+        name: `${firstName} ${lastName}`,
+        firstName: firstName,
+        reports: reports,
+        lastName: lastName,
+        title: title,
+        active: active,
+        phone: phone,
+        hireDate: new Date(hireDate)
+    };
+    return emp;
+}
+
+export const employeeTitles = ["Chief Executive Officer",
+        "Chief Technical Officer",
+        "VP, Engineering",
+        "Team Lead",
+        "Software Architect",
+        "Director, Engineering",
+        "QA Architect", "Chief Product Officer"];
+
+export function makeEmployees(data: any[] , containsField: string
+): Employee[] {
+    function test(emp: Employee): Employee {
+        return emp;
+    }
+    function mapTree(nodes: any[]): void {
+        if (!nodes ) {
+            return;
+        }
+        for (let obj of nodes ) {
+            obj.hireDate = new Date(obj.hireDate);
+            obj = obj;
+            test(obj);
+            mapTree(obj[containsField]);
+        }
+    }
+
+    mapTree([data[0]]);
+    return data;
+}
+
+@Injectable()
+export class EmployeesService extends BehaviorSubject<any> {
+
+    private dataHier: any[] = makeEmployees(employeeHier, 'reports');
+
+    private data: any[] = makeFlatData(this.dataHier, 'id', 'managerId', 'reports');
+
+    constructor() {
+        super(null);
+    }
+
+    public readStatic(query: any = {skip: 0, take: 10}, flat: boolean = true): Array<any> {
+        if (flat) {
+            let piece = this.data.slice(query.skip, query.take);
+            return piece;
+        } else {
+            return this.dataHier;
+        }
+    }
+    public readAsync(query: any = {skip: 0, take: 10}, flat: boolean = true, time: number = 0 ): Observable<any> {
+        if (flat) {
+            let piece = this.data.slice(query.skip, query.take);
+            return of(piece).pipe(delay(time));
+        } else {
+            return of(this.dataHier).pipe(delay(time));
+        }
+    }
+
+    public save(item: any, isNew: boolean, time: number = 0, error: boolean = false): Observable<any> {
```

Comment:
The meaning of `time` and `error` as arguments is not clear. It should denote that these are faked behaviors for the resulting observable.